pkg/cmd/deployment/list/list.go

```diff
@@ -0,0 +1,232 @@
+package list
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/MakeNowJust/heredoc"
+	"github.com/cli/cli/v2/api"
+	"github.com/cli/cli/v2/internal/ghrepo"
+	"github.com/cli/cli/v2/internal/tableprinter"
+	"github.com/cli/cli/v2/pkg/cmdutil"
+	"github.com/cli/cli/v2/pkg/iostreams"
+	"github.com/shurcooL/githubv4"
+	"github.com/spf13/cobra"
+)
+
+type listOptions struct {
+	BaseRepo   func() (ghrepo.Interface, error)
+	HttpClient func() (*http.Client, error)
+	IO         *iostreams.IOStreams
+	Now        time.Time
+
+	Limit       int
+	Environment string
+}
+
+type deployment struct {
+	ID          int
+	Ref         *string
+	Environment string
+	Status      *string
+	CreatedAt   time.Time
+}
+
+type deploymentsQuery struct {
+	Repository struct {
+		Deployments struct {
+			Nodes []struct {
+				DatabaseID  int
+				Environment string
+				Ref         *struct {
+					Name string
+				}
+				LatestStatus *struct {
+					State     githubv4.DeploymentStatusState
+					CreatedAt time.Time
+				}
+				CreatedAt time.Time
+			}
+			PageInfo struct {
+				HasNextPage bool
+				EndCursor   string
+			}
+		} `graphql:"deployments(first: $per_page, environments: $environments, after: $endCursor, orderBy: {field: CREATED_AT, direction: DESC})"`
+	} `graphql:"repository(owner: $owner, name: $name)"`
+}
+
+func NewCmdList(f *cmdutil.Factory, runF func(*listOptions) error) *cobra.Command {
+	opts := listOptions{
+		HttpClient: f.HttpClient,
+		IO:         f.IOStreams,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "list",
+		Short: "List deployments",
+		Long: heredoc.Docf(`
+			List deployments.
+		`, "`"),
+		Example: heredoc.Doc(`
+			# List deployments in the current repository
+			$ gh deployment list
+
+			# List deployments in a specific repository
+			$ gh deployment list --repo owner/repo
+
+			# List deployments in a specific environment
+			$ gh deployment list --environment production
+
+			# List more than 10 deployments
+			$ gh deployment list --limit 100
+		`),
+		Args: cobra.NoArgs,
+		RunE: func(c *cobra.Command, args []string) error {
+			opts.BaseRepo = f.BaseRepo
+
+			if runF != nil {
+				return runF(&opts)
+			}
+			return listRun(&opts)
+		},
+	}
+
+	cmd.Flags().IntVarP(&opts.Limit, "limit", "L", 10, "Maximum number of deployments to fetch")
```

Comment:
**Question**

I don't think it needs to be as part of this PR but would you agree that it might make sense to have flags for:
 * `sha`
 * `ref`
 * `task`

As documented https://docs.github.com/en/rest/deployments/deployments?apiVersion=2022-11-28#list-deployments

Reply:
Yep, you totally could.

I kept to `environment` only as that aligns to what you can filter by in the GitHub UI (in the current and beta versions of it).